File: Python/Searching and Sorting/17_sort_by_set_bit_count.py
"""
Idea:
sorting the array using lambda function
bin(x) converts string to its binary form ( 5 -> '0b101'
bin(x)[2:0]-removing 1st 2 chars ('0b')
count('1')- counts no. of 1 in string
"""


class Solution:
    # Function to sort the array by the count of set bits in each element
    def sortBySetBitCount(self, arr, n):
        # Sorting the array based on the count of set bits in each element
        arr.sort(key=lambda x: -bin(x)[2:].count('1'))


def main():
    T = int(input())

    while T > 0:
        n = int(input())
        arr = [int(x) for x in input().strip().split()]
        ob = Solution()
        ob.sortBySetBitCount(arr, n)
        print(*arr)

        T -= 1


if __name__ == "__main__":
    main()

File: C++/Searching and Sorting/17_sort_by_set_bit_count.cpp
/*
Idea:
Using sort with custom comparator
stable_sort- is used to maintain the relative order of elements with the same count of set bits
__builtin_popcount- counts the number of set bits in the binary representation of integer
*/

#include <bits/stdc++.h>
using namespace std;


class Solution {
public:
    // custom comparator of std::sort
    static int cmp(int a, int b)
    {
        int count1 = __builtin_popcount(a);
        int count2 = __builtin_popcount(b);

        // this takes care of the stability of sorting algorithm too
        if (count1 <= count2)
            return false;
        return true;
    }

    // Function to sort according to bit count using  std::sort
    void sortBySetBitCount(int arr[], int n)
    {
        stable_sort(arr, arr + n, cmp);
    }
};


int main()
{
    int t;
    cin >> t;
    while (t--) {
        int n;
        cin >> n;
        int arr[n];
        for (int i = 0;i < n;i++) {
            cin >> arr[i];
        }
        Solution ob;
        ob.sortBySetBitCount(arr, n);
        for (int i = 0; i < n; i++)
            cout << arr[i] << " ";
        cout << endl;
    }
    return 0;
}
